cpp/src/io/utilities/data_sink.hpp

```diff
@@ -65,7 +79,41 @@ class data_sink {
    *
    * @return void
    **/
-  virtual void write(void const* data, size_t size) = 0;
+  virtual void host_write(void const* data, size_t size) = 0;
+
+  /**
+   * @brief Whether or not this sink supports writing from gpu memory addresses.
+   * 
+   * Internal to some of the file format writers, we have code that does things like
+   * 
+   * tmp_buffer = alloc_temp_buffer();
+   * cudaMemcpy(tmp_buffer, device_buffer, size);
+   * sink->write(tmp_buffer, size);
+   * 
+   * In the case where the sink type is itself a memory buffered write, this ends up
+   * being effectively a second memcpy.  So a useful optimization for a "smart" 
+   * custom data_sink is to do it's own internal management of the movement 
+   * of data between cpu and gpu; turning the internals of the writer into simply
+   * 
+   * sink->device_write(device_buffer, size)
+   * 
+   * If this function returns true, the data_sink will receive calls to device_write() 
+   * instead of write() when possible.  However, it is still possible to receive
+   * write() calls as well.
+   *
+   * @return bool If this writer supports device_write() calls.
+   **/
+  virtual bool supports_device_write() = 0;
+
+  /**
+   * @brief Append the buffer content to the sink from a gpu address
+   *
+   * @param[in] data Pointer to the buffer to be written into the sink object
+   * @param[in] size Number of bytes to write
+   *
+   * @return void
+   **/
+  virtual void device_write(void const* gpu_data, size_t size, cudaStream_t stream) = 0;
```

Comment:
I'd like to avoid this because (for example) in the parquet case, it allocates a worst-case sized buffer and uses it multiple times for sink->write() calls within one writer::write() call.  If we provided a default implementation, it would have to allocate and free on the spot (or do some kind of hinting/resizing thing which seems a little heavyweight for a default implementation) every time which would be pretty bad, performance-wise.

Reply:
How about a default implementation that throws then? Makes it consistent with the supports...() implementation and removes duplicated code.

Reply:
Hold on. This is a pure virtual function, which means any concrete derived class must implement it. But `supports_device_write`  returning false implies it's not there. So why not instead of pure virtual, just make this virtual with a default implementation that does nothing?

That way one can create a derived sink class that doesn't support device_write without having to implement an override for this function.

Reply:
I suggested a default implementation that throws, since the function should not be called (probably won't have the expected effect if it does nothing). It is what classes that don't support device writes do now. @harrism what do you think about this option?

Reply:
Sounds good.

Reply:
Agree.  Change made.